ENH: Addition of optional visitor-functions in extract_text()

srogmann opened the pull request:
This request adds optional visitor-callbacks in `extract_text()`.

`_extract_text()` calls these visitor-methods while scanning the text-objects of a `page`. So one can analyze the operations in the page and the positions of the texts.

`tests/test_page.py` extracts the texts of labels in a Figure and serves as an example how to use this enhancement.

MartinThoma commented:
Thank you for the contribution :heart: 

I didn't expect that we could get text-tokens and their positions in the document in such a rather easy extension. Nice!

I still need to think about this PR / check if there is a performance impact and look at it from a maintenance perspective.

In the meantime, would you mind running `black .`? You need `pip install black`; it's a code-formatter that fixes all of the Flake8 issues.

srogmann commented:
I executed black, it reformatted my changes in _page.py and test_page.py :-).

pubpub-zz commented:
@srogmann,
some extra parameters to be returned to the functions could be useful for some filterig : BaseFont Name and Size (rescaled to the page) ; this would be useful for title extraction for example

srogmann commented:
@pubpub-zz 
I added the font-dictionary and the font-size in the text-visitor-function. I added the reference to the font-dictionary instead of the BaseFont because I didn't know what might be of further interest. 

        def print_visi(text, cm_matrix, tm_matrix, font_dict, font_size):
            if text.strip() != "":
                listTexts.append(
                    PositionedText(
                        text, tm_matrix[4], tm_matrix[5], font_dict, font_size
                    )
                )
[...]

    # Check the fonts. We check: /F2 9.96 Tf [...] [(Dat)-2(e)] TJ
    textDatOfDate = listRows[0][0][0]
    assert textDatOfDate.font_dict is not None
    assert textDatOfDate.font_dict["/Name"] == "/F2"
    assert textDatOfDate.font_dict["/BaseFont"] == "/Arial,Bold"
    assert textDatOfDate.font_dict["/Encoding"] == "/WinAnsiEncoding"
    assert textDatOfDate.font_size == 9.96`

srogmann commented:
@pubpub-zz 
One could add helper classes like PositionedText to support parsing of formatted texts. I used tests/test_page.py as some kind of inkubator ;-).

` assert textDat.get_base_font() == "/Arial,Bold"`

MartinThoma commented:
@srogmann What is the state of this PR? Do you need help to resolve the merge conflicts / the failing test?

Besides those, is the PR ready in your opinion?

srogmann commented:
@MartinThoma In my opinion the PR was ready 22 days ago. I will have a look at the current state.

MartinThoma commented:
I'm sorry for the delay; I thought there still was something to be done :see_no_evil:

srogmann commented:
@MartinThoma  I merged and resolved conflicts. The PR should work. You may have a look at it.

Each change of the output-result in _extract_text requires a visitor-call in _extract_text:
```
                    if visitor_text is not None:
                        visitor_text(text, cm_matrix, tm_matrix, cmap[3], font_size)
```
There are quite a lot of variables describing the current state, cmap[3] contains the font-dictionary. Perhaps in future there will be one object describing the state.

In tests/test_page.py there are some functions which might be of interest in PyPDF2 or pdfly in future to support using a visitor.
For example one might try to create a svg file:

```
    def exportSvgFile(listTexts, listRects, fileName):
        import svgwrite

        dwg = svgwrite.Drawing(fileName, profile="tiny")
        color = svgwrite.rgb(255, 0, 0, "%")
        for r in listRects:
            dwg.add(dwg.rect((r.x, r.y), (r.w, r.h), stroke=color, fill_opacity=0.05))
        for t in listTexts:
           dwg.add(dwg.text(t.text, insert=(t.x, t.y), fill="blue"))
        dwg.save()
```

MartinThoma commented on tests/test_page.py:
```diff
@@ -331,6 +331,258 @@ def test_extract_text_operator_t_star():  # L1266, L1267
+    # Test 1: We test the analysis of page 7 "2.1 LRS model".
+    reader = PdfReader(RESOURCE_ROOT / "GeoBase_NHNC1_Data_Model_UML_EN.pdf")
+    pageLrsModel = reader.pages[6]
+    # We ignore the invisible large rectangles.
```
```suggestion

    # We ignore the invisible large rectangles.
```

MartinThoma commented on tests/test_page.py:
```diff
@@ -331,6 +331,258 @@ def test_extract_text_operator_t_star():  # L1266, L1267
+    # Test 2: Parse table "REVISION HISTORY" on page 3.
+    pageRevisions = reader.pages[2]
+    # We ignore the second table, therefore: r.y > 350
+    def filterFirstTable(r):
```
```suggestion

    def filterFirstTable(r):
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1167,13 +1167,28 @@ def _debug_for_extract(self) -> str:  # pragma: no cover
         orientations: Tuple[int, ...] = (0, 90, 180, 270),
         space_width: float = 200.0,
         content_key: Optional[str] = PG.CONTENTS,
+        visitor_operand_before: Optional = None,
+        visitor_operand_after: Optional = None,
+        visitor_text: Optional = None,
```
```suggestion
        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None] = None,
        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None] = None,
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1565,6 +1666,9 @@ def extract_text(
         TJ_sep: str = None,
         orientations: Union[int, Tuple[int, ...]] = (0, 90, 180, 270),
         space_width: float = 200.0,
+        visitor_operand_before: Optional = None,
+        visitor_operand_after: Optional = None,
+        visitor_text: Optional = None,
```
```suggestion
        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None] = None,
        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None] = None,
```

srogmann commented:
```
-        visitor_operand_before: Optional = None,
-        visitor_operand_after: Optional = None,
-        visitor_text: Optional = None,
+        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
+        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None] = None,
+        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None] = None,
```
Fine :-). I didn't know how to do that in Python.

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1254,13 +1254,28 @@ def _debug_for_extract(self) -> str:  # pragma: no cover
         orientations: Tuple[int, ...] = (0, 90, 180, 270),
         space_width: float = 200.0,
         content_key: Optional[str] = PG.CONTENTS,
+        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
+        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None] = None,
+        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None] = None,
```
```suggestion
        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None]] = None,
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1652,6 +1753,9 @@ def extract_text(
         TJ_sep: str = None,
         orientations: Union[int, Tuple[int, ...]] = (0, 90, 180, 270),
         space_width: float = 200.0,
+        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
+        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None] = None,
+        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None] = None,
```
```suggestion
        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None]] = None,
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1254,13 +1254,28 @@ def _debug_for_extract(self) -> str:  # pragma: no cover
             out += "No Font\n"
         return out
 
+    def _visit_operand_empty(op, args, cm_matrix, tm_matrix):
+        """
+        Dummy-implementation of an operand-visitor called by _extract_text.
+        """
+        pass
+
+    def _visit_text_empty(text, cm_matrix, tm_matrix):
+        """
+        Dummy-implementation of a text-visitor called by _extract_text.
+        """
+        pass
+
```
```suggestion
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1708,14 +1813,24 @@ def extract_text(
         xform: EncodedStreamObject,
         orientations: Tuple[int, ...] = (0, 90, 270, 360),
         space_width: float = 200.0,
+        visitor_operand_before: Optional = None,
+        visitor_operand_after: Optional = None,
+        visitor_text: Optional = None,
```
```suggestion
        visitor_operand_before: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_operand_after: Optional[Callable[[Any, Any, Any, Any], None]] = None,
        visitor_text: Optional[Callable[[Any, Any, Any, Any, Any], None]] = None,
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1301,11 +1318,14 @@ def _extract_text(
                 cmaps[f] = build_char_map(f, space_width, obj)
-        cmap: Tuple[Union[str, Dict[int, str]], Dict[str, str], str] = (
+        cmap: Tuple[
+            Union[str, Dict[int, str]], Dict[str, str], str, DictionaryObject
```
```suggestion
            Union[str, Dict[int, str]], Dict[str, str], str, Optional[DictionaryObject]
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1622,17 +1694,30 @@ def process_operation(operator: bytes, operands: List) -> None:
                     if xobj[operands[0]]["/Subtype"] != "/Image":  # type: ignore
                         # output += text
-                        text = self.extract_xform_text(xobj[operands[0]], orientations, space_width)  # type: ignore
+                        text = self.extract_xform_text(
+                            xobj[operands[0]],
+                            orientations,
+                            space_width,
+                            visitor_operand_before,
+                            visitor_operand_after,
+                            visitor_text,
+                        )  # type: ignore
```
```suggestion
                        text = self.extract_xform_text(
                            xobj[operands[0]],  # type: ignore
                            orientations,
                            space_width,
                            visitor_operand_before,
                            visitor_operand_after,
                            visitor_text,
                        )
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1430,21 +1456,29 @@ def process_operation(operator: bytes, operands: List) -> None:
+                        charMapTuple[2],
+                        charMapTuple[3],
                         operands[0],
+                        charMapTuple[4],
```
mypy complains:

> error: Tuple index out of range  [misc]

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1301,11 +1318,14 @@ def _extract_text(
         if "/Font" in resources_dict:
             for f in cast(DictionaryObject, resources_dict["/Font"]):
                 cmaps[f] = build_char_map(f, space_width, obj)
```
mypy complains:

> error: Incompatible types in assignment (expression has type "Tuple[str, float, Union[str, Dict[int, str]], Dict[Any, Any], DictionaryObject]", target has type "Tuple[str, float, Union[str, Dict[int, str]], Dict[str, str]]")  [assignment]

MartinThoma commented:
@srogmann I've fixed a couple of type annotation issues. Sadly, I can only adjust the lines that you have already touched. Within `_page.py` you need to change

```python
        cmaps: Dict[
            str, Tuple[str, float, Union[str, Dict[int, str]], Dict[str, str]]
        ] = {}
```

to
```python
        cmaps: Dict[
            str,
            Tuple[
                str, float, Union[str, Dict[int, str]], Dict[str, str], DictionaryObject
            ],
        ] = {}
```

srogmann commented:
@MartinThoma I added DictionaryObject in cmaps (my last commit-comment is wrong).

MartinThoma commented:
@srogmann Thank you for your contribution! If you want, I can add you to https://pypdf2.readthedocs.io/en/latest/meta/CONTRIBUTORS.html :-)

MartinThoma commented:
> There are quite a lot of variables describing the current state, cmap[3] contains the font-dictionary. Perhaps in future there will be one object describing the state.

Yes, I think that would make sense. Also getting rid of the use of global / non-local variables and passing data explicitly around might help.

MartinThoma reviewed (approved):
Looks good to me!

MartinThoma commented:
@pubpub-zz Did you have a look? What do you think about the changes?

If you're good with them as well, I would merge + release :-)

pubpub-zz commented:
This sounds good. I had some request earlier that have been fullfiled.
I agree that it is time to release it for user feedbacks

MartinThoma commented:
@srogmann Very nice work :partying_face: 

I think this has a lot of potential, but it's also pretty hard to use. We will need to add documentation and typical examples. I'm super excited to see how people will use it :tada:

srogmann commented:
@MartinThoma  Thanks for merging!

An additation to CONTRIBUTORS.html would be fine.

> I think this has a lot of potential, but it's also pretty hard to use. We will need to add documentation and typical examples. 

In tests/test/page.py there a two util-classes PositionedText and Rectangle. After renaming they might be useful when one wants to write an own visitor. Documentation and typical examples would be nice. But docs/user/ is contained in the repository, too, so I can think about another pull-request containing some documentation and examples (e.g. the util-classes mentioned and a sample to extract tables or to ignore page-headers).
